### PR TITLE
BIGC-658 Basic user data management

### PR DESCRIPTION
**Pull request by sofib**
# Pull Request Template

## Description

Please include a summary of the change and which issue is fixed. Please also include relevant motivation and context. List any dependencies that are required for this change.

## Type of change

- [ ] Bug fix
- [x] New feature
- [ ] This change requires a documentation update

## Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [x] I updated the CHANGELOG.md

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -5,6 +5,7 @@ All notable changes to this project will be documented in this file.
 The format is based on [Keep a Changelog](http://keepachangelog.com/) and this project adheres to [Semantic Versioning](http://semver.org/).
 
 ## [Unreleased]
+- basic used data management
```
Comment (chrisEff) [inline CHANGELOG.md]: typo: used
```suggestion
- basic user data management
```

File: extension/lib/bigcommerce/CustomerRepository.js [hunk 2]
```diff
@@ -68,11 +70,28 @@ class BigCommerceCustomerRepository {
...
+   */
+  async getCustomerById (customerId) {
+    const uri = `/customers/${customerId}`
+    const customer = await this.apiClientV2.get(uri)
 
-    if (!customers || customers.length < 1) return null
+    if (!customer) return null
 
-    return customers[0]
+    return customer
```
Comment (chrisEff) [inline extension/lib/bigcommerce/CustomerRepository.js]:
```suggestion
    return customer || null
```

File: extension/lib/bigcommerce/CustomerRepository.js [hunk 3]
```diff
@@ -68,11 +70,28 @@ class BigCommerceCustomerRepository {
...
+  /**
+   * @param {number} customerId
+   * @returns {Promise<Object>}
+   */
+  async getCustomerById (customerId) {
+    const uri = `/customers/${customerId}`
+    const customer = await this.apiClientV2.get(uri)
 
-    if (!customers || customers.length < 1) return null
+    if (!customer) return null
```
Comment (chrisEff) [inline extension/lib/bigcommerce/CustomerRepository.js]:
```suggestion
```

File: .travis.yml [hunk 4]
```diff
@@ -25,7 +25,7 @@ matrix:
       script:
       - if [[ "$INTEGRATION_CREDENTIALS" ]]; then echo "${INTEGRATION_CREDENTIALS}" > .integration-credentials.js; fi
```
Comment (chrisEff) [inline .travis.yml]: integration-credentials should be in the extension directory (like in the bc-cart extension)
```suggestion
      - if [[ "$INTEGRATION_CREDENTIALS" ]]; then echo "${INTEGRATION_CREDENTIALS}" > extension/.integration-credentials.js; fi
```

File: extension/test/integration/lib/bigcommerce/CustomerRepository-test.js [hunk 5]
```diff
@@ -0,0 +1,66 @@
+const assert = require('assert')
+const chai = require('chai')
+chai.use(require('chai-as-promised')).should()
+const BigCommerceCustomerRepository = require('../../../../lib/bigcommerce/CustomerRepository')
+const BigCommerceRequestClientError = require('../../../../lib/bigcommerce/customer/request/ClientError')
+const { clientId, accessToken, storeHash } = require('../../../../../.integration-credentials')
```
Comment (chrisEff) [inline extension/test/integration/lib/bigcommerce/CustomerRepository-test.js]: integration-credentials should be in the extension directory (like in the bc-cart extension)
```suggestion
const { clientId, accessToken, storeHash } = require('../../../../.integration-credentials')
```

File: extension/lib/bigcommerce/CustomerRepository.js [hunk 6]
```diff
@@ -1,8 +1,10 @@
 const jwtDecoder = require('jwt-simple')
+const BigCommerce = require('node-bigcommerce')
+
 const BigCommerceCustomerTokenInvalidError = require('./customer/jwt/TokenInvalidError')
 const BigCommerceCustomerTokenUnverifiedError = require('./customer/jwt/TokenInvalidError')
```
Comment (chrisEff) [inline extension/lib/bigcommerce/CustomerRepository.js]: (I know this is not part of you change, but) it seems like the wrong error is being imported here
```suggestion
const BigCommerceCustomerTokenUnverifiedError = require('./customer/jwt/TokenUnverifiedError')
```

File: extension/lib/bigcommerce/CustomerRepository.js [hunk 7]
```diff
@@ -98,6 +117,23 @@ class BigCommerceCustomerRepository {
...
+
+  /**
+   * Update customer entity on bigc side.
+   * ATM custom fields (form_fields) are not supported see https://support.bigcommerce.com/s/question/0D51B000046LK53SAG/how-to-update-custom-customer-fields-using-api
+   * @param {number} customerId
+   * @param {BigCommerceCustomerRequest} customer
+   * @return {Promise<void>}
+   */
+  async update (customerId, customer) {
+    console.error('updating', customer)
```
Comment (chrisEff) [inline extension/lib/bigcommerce/CustomerRepository.js]: I guess you didn't mean to commit that... :-)
```suggestion
```

File: extension/lib/bigcommerce/customer/shapeRequestError.js [hunk 8]
```diff
@@ -0,0 +1,28 @@
...
+    return err
+  }
+
+  let parsed
+  try {
+    parsed = JSON.parse(errorMessageMatch[1])
+  } catch (unparseable) {
+    return new BigCommerceRequestUnparseableMessageError('unable to parse')
+  }
+
```
Comment (chrisEff) [inline extension/lib/bigcommerce/customer/shapeRequestError.js:17]: If you'd move everything below this line into the try block, you wouldn't have the problem of using the `parsed` variable in different scopes. Thus, you could define it with `const`.

File: extension/lib/bigcommerce/customer/shapeRequestError.js [hunk 9]
```diff
@@ -0,0 +1,28 @@
+const BigCommerceRequestClientError = require('./request/ClientError')
+const BigCommerceRequestUnparseableMessageError = require('./request/UnparsableMessageError')
+
+module.exports = function shapeRequestError (err) {
+  // Try parsing the (potential) underlying api error
+  const errorMessageMatch = err.message.match(/({.+})/)
```
Comment (chrisEff) [inline extension/lib/bigcommerce/customer/shapeRequestError.js]: You can omit the parenthesis here...
```suggestion
  const errorMessageMatch = err.message.match(/{.+}/)
```

File: extension/lib/bigcommerce/customer/shapeRequestError.js [hunk 10]
```diff
@@ -0,0 +1,28 @@
...
+  if (!errorMessageMatch) {
+    return err
+  }
+
+  let parsed
+  try {
+    parsed = JSON.parse(errorMessageMatch[1])
```
Comment (chrisEff) [inline extension/lib/bigcommerce/customer/shapeRequestError.js]: ... but then you have to use the 0th index here.
```suggestion
    parsed = JSON.parse(errorMessageMatch[0])
```

**Review by chrisEff — CHANGES_REQUESTED** [review]
I suggested a few minor changes.

File: extension/lib/bigcommerce/customer/shapeRequestError.js (hunk 8 above)
Comment (sofib) [reply extension/lib/bigcommerce/customer/shapeRequestError.js:17]: I think this is a matter of style, and there are some benefits to it that might not be obvious at the fist glance - for instance when we want to have everything inside try/catch block it is often the case that we have the need to throw a custom error - that will cause the error to be caught by the same try/catch block which is not good practice. From my point of view at least.

Comment (chrisEff) [reply extension/lib/bigcommerce/customer/shapeRequestError.js:17]: ok

**Review by chrisEff — APPROVED** [review]

**Pull request merged**